source/s3_client.c

```diff
@@ -1193,27 +1194,9 @@ static void s_s3_client_process_work_default(struct aws_s3_client *client) {
     uint32_t num_requests_queued =
         aws_s3_client_queue_requests_threaded(client, &client->synced_data.prepared_requests, false);
 
-    {
-        int sub_result = aws_sub_u32_checked(
-            client->threaded_data.num_requests_being_prepared,
-            num_requests_queued,
-            &client->threaded_data.num_requests_being_prepared);
-
-        AWS_ASSERT(sub_result == AWS_OP_SUCCESS);
-        (void)sub_result;
-    }
-
-    {
-        int sub_result = aws_sub_u32_checked(
-            client->threaded_data.num_requests_being_prepared,
-            client->synced_data.num_failed_prepare_requests,
-            &client->threaded_data.num_requests_being_prepared);
-
-        client->synced_data.num_failed_prepare_requests = 0;
-
-        AWS_ASSERT(sub_result == AWS_OP_SUCCESS);
-        (void)sub_result;
-    }
+    aws_atomic_fetch_sub(&client->stats.num_requests_being_prepared, num_requests_queued);
+    aws_atomic_fetch_sub(&client->stats.num_requests_being_prepared, client->synced_data.num_failed_prepare_requests);
```

Comment:
this is a big change.
We've never depended on the `stats` for behavior before, right?

I believe they've just been used for performance analysis so far. I'm not sure we've been super careful about ensuring their correctness. That's why they're atomics and not properly controlled by a lock, we didn't super care if we printed all the stats and the numbers didn't add up.

Reply:
semi related to this: should we consider just simplifying the current logic? 

we currently increment being_prepared count when we prep requests, then in prep_done callback we push prepped request to queue and update failed numbers and then in the scheduling loop we try to rectify what the number of being prepared should be. all of this happens in critical sections. 

should the logic to dec being_prepared just live in prep_done callback? i.e. always inc when kicking off prep and always dec when prep is done

Reply:
yeah, chatted with Mike, the logic will be simpler if just keep the no-op requests in the queue, and clean them up when the client literate through the queue.

The callback can be invoked from another thread, if we decrease counter from the callback, we will need a lock around it, and increase the complexity of the code.

Reply:
callback already has a lock on client. but sure, im ok with removing noops from queue later as well

Reply:
yeah, callback has a lock, but we don't have lock to increase the number. https://github.com/awslabs/aws-c-s3/blob/main/source/s3_client.c#L1387